cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,1280 @@
+
+    thrust::reduce_by_key(handle.get_thrust_policy(),
+                          get_dataframe_buffer_begin(vertex_pair_buffer_tmp),
+                          get_dataframe_buffer_end(vertex_pair_buffer_tmp),
+                          num_triangles_tmp.begin(),
+                          get_dataframe_buffer_begin(vertex_pair_buffer),
+                          num_triangles.begin(),
+                          thrust::equal_to<thrust::tuple<vertex_t, vertex_t>>{});
+
+    auto edges_to_num_triangles = thrust::make_zip_iterator(
+      get_dataframe_buffer_begin(vertex_pair_buffer), num_triangles.begin());
+
+    // 'invalid_edge_first' marks the beginning of the edges to be removed
+    auto invalid_edge_first =
+      thrust::stable_partition(handle.get_thrust_policy(),
+                               edges_to_num_triangles,
+                               edges_to_num_triangles + num_triangles.size(),
+                               [k] __device__(auto e) {
+                                 auto num_triangles = thrust::get<1>(e);
+                                 auto is_in_k_truss = num_triangles == 2;
+                                 return num_triangles > k;  // FIXME (k-2) * 2
+                               });
+
+    size_t num_invalid_edges{0};
+    num_invalid_edges = static_cast<size_t>(
+      thrust::distance(invalid_edge_first, edges_to_num_triangles + num_triangles.size()));
+
+    size_t invalid_edge_start_idx{0};
+    invalid_edge_start_idx =
+      static_cast<size_t>(thrust::distance(edges_to_num_triangles, invalid_edge_first));
+
+    // copy invalid edges
+    auto invalid_edges_buffer = allocate_dataframe_buffer<thrust::tuple<vertex_t, vertex_t>>(
+      num_invalid_edges, handle.get_stream());
+
+    thrust::copy(handle.get_thrust_policy(),
+                 get_dataframe_buffer_begin(vertex_pair_buffer) + invalid_edge_start_idx,
+                 get_dataframe_buffer_end(vertex_pair_buffer),
+                 get_dataframe_buffer_begin(invalid_edges_buffer));
+
+    while (num_invalid_edges != 0) {  // unroll and remove/mask edges
+
+      // case 2: unroll (q, r)
+
+      auto incoming_vertex_pairs_begin =
+        thrust::make_zip_iterator(edgelist_dsts.begin(), edgelist_srcs.begin());
+
+      vertex_t num_incomming_vertex_pairs = edgelist_srcs.size();
+
+      // Sort the 'incoming_vertex_pairs_begin' by 'dst'
+      thrust::sort(
+        handle.get_thrust_policy(),
+        incoming_vertex_pairs_begin,
+        incoming_vertex_pairs_begin + edgelist_srcs.size());  // FIXME: No need to partition
+
+      // For each (q, r) edge to unroll, find the incoming edges to 'r' let's say from 'p' and
+      // create the pair (p, q)
+      rmm::device_uvector<int> prefix_sum(edgelist_srcs.size() + 1, handle.get_stream());
+      thrust::tabulate(
+        handle.get_thrust_policy(),
+        prefix_sum.begin(),
+        prefix_sum.end(),
```

Comment:
You really don't need to iterate over the entire set of edges.

You just need to iterate over the edges to be deleted. And unroll their contributions to the edges to be remained.

Iterate all the edges to be deleted, and consider the case they are p,q edges. For each r, decrement triangle counts by 1 for (p, r) and (q, r) edges (if they are in the remaining edge list).

Then, consider the (p, r) & (q, r) case (iterate over only the edges to be deleted). Need to do the same, but you need to consider the case that their contribution is already considered in considering the (p,q) case as we discussed in the past.

The code below looks like doing way more work than what is strictly necessary.

Reply:
> You just need to iterate over the edges to be deleted. And unroll their contributions to the edges to be remained

Right that was a mistake and it is now resolved